core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -57,6 +57,7 @@ class LogSegment private[log] (val log: FileRecords,
                                val lazyOffsetIndex: LazyIndex[OffsetIndex],
                                val lazyTimeIndex: LazyIndex[TimeIndex],
                                val txnIndex: TransactionIndex,
+                               val producerStateSnapshot: File,
```

Comment:
Hmm, it seems that we are missing the handling of producerStateSnapshot in a few methods like changeFileSuffixes(), updateDir(), and potentially flush().

Reply:
Hmm I think if we want to treat producer snapshot files like other index files, it seems we may need a different approach. In this PR I was relying on `cleanupOrphanSnapshotFiles` to avoid having to deal with `changeFileSuffixes` and `updateDir`.

Reply:
Hmm, by putting producerStateSnapshot inside LogSegment, it seems to suggest that producerStateSnapshot should be managed as other files belonging to a segment?

Reply:
I would agree.

Reply:
That sounds good, I was just saying the approach taken in this PR will have to be rethought depending on what degree we want the `LogSegment` to manage producer state snapshots versus the `Log`/`ProducerStateManager`. I can think of two options which can be taken to improve the ownership here a bit. 

1. The `Log` continues to manage producer state snapshots via the `ProducerStateManager` and the logic for deletion is lifted back into the log layer. This means wherever we try to delete segments with `LogSegment.deleteIfExists` we also delete the corresponding producer state snapshot via a call to the `ProducerStateManager`. I think this most closely resembles the current state in trunk.

2. The `LogSegment` and `Log`/`ProducerStateManager` share ownership of the producer state snapshot with only the `LogSegment` owning deletion and access going through the `ProducerStateManager`. This may allow us to get rid of the producer state snapshot specific truncation logic in `ProducerStateManager`. The question here is what to do about producer state snapshot files which don't match a corresponding segments base offset. This may not be too bad as I think in all cases they would just be associated with the active segment.

Reply:
I see. The tricky thing is that the last snapshot doesn't have a corresponding log segment and needs to be managed separately outside of the LogSegment. However, for snapshots that have an associated segment, managing the snapshot file as part of the log segment seems convenient.